File: src/modules/algorithm/seed/algorithms/stack/languages/python/linkedlist-solution.py
class Node:
    def __init__(self, data: int):
        """
        Initializes a node with data and a next pointer.
        """
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        """
        Initializes an empty stack using a linked list.
        """
        self.top = None

    def push(self, value: int) -> None:
        """
        Adds an element to the top of the stack.
        """
        new_node = Node(value)
        new_node.next = self.top
        self.top = new_node

    def pop(self) -> int | None:
        """
        Removes and returns the top element from the stack.
        Returns None if the stack is empty.
        """
        if self.is_empty():
            return None
        value = self.top.data
        self.top = self.top.next
        return value

    def peek(self) -> int | None:
        """
        Returns the top element of the stack without removing it.
        Returns None if the stack is empty.
        """
        if self.is_empty():
            return None
        return self.top.data

    def is_empty(self) -> bool:
        """
        Returns True if the stack is empty, False otherwise.
        """
        return self.top is None 

File: src/modules/algorithm/seed/algorithms/stack/languages/typescript/linkedlist-solution.ts
export interface Stack {
  push(value: number): void;
  pop(): number | undefined;
  peek(): number | undefined;
  isEmpty(): boolean;
}

class Node {
  data: number;
  next: Node | null;

  constructor(data: number) {
    this.data = data;
    this.next = null;
  }
}

export class StackImpl implements Stack {
  private top: Node | null = null;

  push(value: number): void {
    const newNode = new Node(value);
    newNode.next = this.top;
    this.top = newNode;
  }

  pop(): number | undefined {
    if (this.isEmpty()) {
      return undefined;
    }
    const value = this.top!.data;
    this.top = this.top!.next;
    return value;
  }

  peek(): number | undefined {
    if (this.isEmpty()) {
      return undefined;
    }
    return this.top!.data;
  }

  isEmpty(): boolean {
    return this.top === null;
  }
}
